samza-core/src/main/java/org/apache/samza/zk/ZkUtils.java

```diff
@@ -110,17 +109,92 @@ public ZkKeyBuilder getKeyBuilder() {
    * @return String representing the absolute ephemeralPath of this client in the current session
    */
   public synchronized String registerProcessorAndGetId(final ProcessorData data) {
+    String processorId = data.getProcessorId();
     if (ephemeralPath == null) {
-      ephemeralPath =
-          zkClient.createEphemeralSequential(
-              keyBuilder.getProcessorsPath() + "/", data.toString());
-
-      LOG.info("newly generated path for " + data +  " is " +  ephemeralPath);
-      return ephemeralPath;
+      /**
+       * If there is an existing processor with same processorId as incoming processorId, throw up and fail.
+       * Caller of this api should handle exception and kill the stream processor.
+       */
+      Preconditions.checkState(getProcessorCount(processorId) == 0,
+                               "Processor: %s is duplicate in the group. Registration failed.", processorId);
+      ephemeralPath = registerProcessor(data);
     } else {
-      LOG.info("existing path for " + data +  " is " +  ephemeralPath);
-      return ephemeralPath;
+      LOG.info("Ephemeral path: {} exists for processor: {} in zookeeper.", ephemeralPath, data);
+    }
+    return ephemeralPath;
+  }
+
+  /**
+   * Registers a processor with zookeeper. Creates a sequential ephemeral node in zookeeper(which will be deleted
+   * on zkClient session timeout) for the processor under processors directory.
+   * @param processorData to register in zookeeper.
+   * @return the sequential ephemeral path created in zookeeper.
+   */
+  String registerProcessor(final ProcessorData processorData) {
+    // Create ephemeral path in zookeeper corresponding to this processor.
+    ephemeralPath = zkClient.createEphemeralSequential(keyBuilder.getProcessorsPath() + "/", processorData.toString());
+    LOG.info("Created ephemeral path: {} for processor: {} in zookeeper.", ephemeralPath, processorData);
+    ProcessorNode processorNode = new ProcessorNode(processorData, ephemeralPath);
+    String processorId = processorData.getProcessorId();
+    // Determine if there are duplicate processors with this.processorId after registration.
+    if (getProcessorCount(processorId) > 1 && shouldStopProcessor(processorNode)) {
```

Comment:
When multiple clients(`C1, C2... Cn`) are trying to register with same processorId (`P1`) at same time, `shouldStopProcessor` picks a client to live and stops rest of them(instead of stoping all).

Current implementation will pick client with lowest `zkProcessorPath` to live and stop others.

This behavior is verified in `TestZkUtils` integration test.